src/Primitives.hs

```diff
@@ -151,65 +151,61 @@ primitiveColumn x@(XObj _ i t) ctx args =
     go = maybe err (\info -> (ctx, Right (XObj (Num IntTy (fromIntegral (infoColumn info))) i t)))
 
 primitiveImplements :: Primitive
-primitiveImplements xobj ctx [x@(XObj (Sym interface@(SymPath _ _) _) _ _), inner@(XObj (Sym impl@(SymPath prefixes name) _) info _)] =
-  let global = contextGlobalEnv ctx
-      def = lookupInEnv impl global
-   in maybe notFound' found def
+primitiveImplements _ ctx [x@(XObj (Sym interface@(SymPath _ _) _) _ _), (XObj (Sym (SymPath prefixes name) _) info _)] =
+  do
+    (maybeInterface, maybeImpl) <- pure ((lookupInEnv interface tyEnv), (lookupInEnv (SymPath modules name) global))
+    case (maybeInterface, maybeImpl) of
+      (_, Nothing) ->
+        if null modules
+          then pure (evalError ctx "Can't set the `implements` meta on a global definition before it is declared." info)
+          else updateMeta (Meta.stub (SymPath modules name)) ctx
+      (Nothing, Just (_, implBinder)) ->
+        (warn >> updateMeta implBinder ctx)
```

Comment:
This use of the monadic operators is really nice.

Reply:
(I think they really shine when the whole flow fits on one line...)

Reply:
Good to keep in mind, thanks!